ls1intum/Artemis
# User Management for GitLab via Artemis

### Checklist

### Motivation and Context


**This PR depends on #1110**
For our GitLab+Jenkins setup we don't have any external user management setup like Jira atm. Because of that, we have to manage all users and groups using the Artemis DB.

### Description

- I extended the user management UI allowing admins to also specify the password and edit the groups a user belongs to Those groups now also get displayed in the user component next to the roles. 
- The server now also provides an `ArtemisInternalAuthenticationProvider` in case `jira` and `ldap` are not activated as a Spring profile. In this case, all users get authenticated and managed using the internal Artemis DB. 
- The GitLab service and VCS interface now also allow Artemis to manage and sync the users and groups. This means that any new user gets also added to GitLab and any updated user (e.g. new password) gets also updated in GitLab. Since we also allow instructors to change the groups of a course after creating it, if there are already exercises in this course, the related repositories and groups in GitLab will also get updated (adding/removing new/old instructors/tutors, upgrading/downgrading instructors/tutors). The same goes for changed groups of a user itself. If a user should not be part of an instructor/tutor group after the update, the related entities in GitLab will also be changed.

### Steps for Testing


0. Message me in order to get an account for the testserver
1. Log in to the Artemis Testserver 2
2. Create some new users
3. Those users should also get created in GitLab
4. Create a new course and a programming exercise
5. The users should have the correct rights for the new programming exercise in GitLab (instructors --> Maintainer, tutors --> Guest)
6. Add some more users (tutors, instructors)
7. Those users should now also be a member of the programming exercise group in GitLab
8. Add some users that do not belong to any instructor/tutor group
9. Change the groups of the course, so that some of the new users should now be an instructor/tutor
10. The groups in GitLab should be updated
11. Change the groups of the course, so that some instructors should now only be tutors and check GitLab
12. Change the groups of the course, so that some tutors should now be instructors and check GitLab
13. Change the groups of the course, so that some users are now longer an instructor or tutor and check GitLab

### Screencast
![gitla_user_management](https://user-images.githubusercontent.com/27979674/72122292-b29f9780-335d-11ea-8e52-9f1657e96de8.gif)
